docs/fpc-registration-new.puml

```diff
@@ -0,0 +1,87 @@
+@startuml
+
+'- force wrapping of longer text
+'skinparam wrapWidth 400
+
+!pragma teoz true
+
+
+title Fabric Private Chaincode - Enclave registry - New approach
+
+actor	Admin order 10
+participant ECC order 20
+participant ECC_Manager order 30
+participant ERCC order 60
+
+note right ERCC
+  ERCC can be **integrated** in ECC or as **seperate** chaincode
+  Integrated:
+    - ERCC and ECC share endorsement policy 
+    - attestations stored ECC namespace
+  Separate:
+    - flexible endorsement policies (ERCC and ECC can have different)
+    - attestations stored in ERCC namespace;
+end note
+
+entity IAS order 50
+collections Orderer order 70
+collections Peers order 80
+
+group instantiation
+  Admin -> ECC : install
+  Admin -> ECC : instantiate (exp.mrenclave)
+  activate ECC
+  note right
+    With new life-cycle mrenclave defined in
+    chaincode def with approve and commit steps
+  end note
+
+  ECC -> ECC : putState(ecc.mrenclave, exp.mrenclave) 
+  ECC --> Admin : success / error
+  deactivate ECC
+end
+
+Admin -> Orderer : submit instantiate transaction
+...when next block ready...
+Orderer -> Peers : broadcast block
+Peers -> Peers : Validation
+
+group enclave creation and attestation generation
```

Comment:
what's the semantics of approve?
According to https://buildmedia.readthedocs.org/media/pdf/hyperledger-fabric/latest/hyperledger-fabric.pdf the chaincode on a channel is defined by a pair name-version, but then... each organization associates a "possibly different" package ID with it -- since such ID contains the hash of the package, then the package itself can be different.
So,
* MRENCLAVE cannot be part of the package, because it could be different
* it should be part of the chaincode definition, but I am not sure that it can be extended
* it could be part of the cc name, but then the infrastructure should enforce that this is equal to the actual enclave's

Reply:
The package id is afaict only a local identifier created as you do a peer lifecycle chaincode install.  That said, packages (and other approve params like plugins) could be different in the approve. However, a subsequent commit will only succeed if sufficient approveformyorg do agree on the same content (i.e., chaincode definition). See January lifecycle play back on https://wiki.hyperledger.org/display/fabric/Playbacks and/or look at lifecycle slide Marcus "extracted" from Ale a while back (lifecycle_pres_2019-113-updated.odp) ..

So that's exactly why you want MRENCLAVE in package (and approve) in it as it exactly the place where each organization explicitly defines the chaincode definition (i.e., the content of the local package as well as the other stuff like plugins) it agrees with.

Reply:
Maybe you're right - docs are a bit confusing about this point.
Alright, in this case, I interpret the two options you mentioned for MRENCLAVE as follows:
* direct: MRENCLAVE is part of the cc definition, which has to be extended (and this might require to modify Fabric).
* indirect: MRENCLAVE is in the cc itself (e,g., in the shim)

Reply:
I guess in principle you could bake in MRENCLAVE into a shim binary but i don't really see the point, at least from an architectural perspective? The shim doesn't really matter (and i'm not sure whether the shim is really in the measurement for the javascript/typescript version or the java one). Also we really want a fixed shim, so it cannot really contain the MRENCLAVE?

That said, where you thinking of the current state before going to a true chaincode type?

Reply:
> I guess in principle you could bake in MRENCLAVE into a shim binary but i don't really see the point, at least from an architectural perspective? The shim doesn't really matter (and i'm not sure whether the shim is really in the measurement for the javascript/typescript version or the java one). Also we really want a fixed shim, so it cannot really contain the MRENCLAVE?

The point is that: if organizations approve the package id, they approve also the MRENCLAVE baked inside.
Thinking about that, arguably this may not be necessary, since the hash of the package should also cover the enclave binary. So it follows that the agreement on the package id, implies agreement on an MRENCLAVE (whatever that is).

> 
> That said, where you thinking of the current state before going to a true chaincode type?

I don't understand the question. Rephrase?

Reply:
> [...] See January lifecycle play back on https://wiki.hyperledger.org/display/fabric/Playbacks and/or look at lifecycle slide Marcus "extracted" from Ale a while back (lifecycle_pres_2019-113-updated.odp) ..

btw: https://hyperledger-fabric.readthedocs.io/en/latest/chaincode4noah.html?highlight=policy gives good discussion of approve/polices and also chaincode definition)

Reply:
> The point is that: if organizations approve the package id, they approve also the MRENCLAVE baked inside.
> Thinking about that, arguably this may not be necessary, since the hash of the package should also cover the enclave binary. So it follows that the agreement on the package id, implies agreement on an MRENCLAVE (whatever that is).

Actually, looking a bit closer at code and docu i'm actually not sure whether the current code of the new lifecycle does bind to the code! I think they did implement now the model where each peer can have a different implementation as long as they agree on other stuff (see link i referenced above). This works fine as long as you only care about integrity (and trust at least as many other peers as required by the endorsement policy quorum). However, in the case of confidentiality, i.e., us, that doesn't work. So we _must_ add MRENCLAVE separatedly!

> > That said, where you thinking of the current state before going to a true chaincode type?
> 
> I don't understand the question. Rephrase?

what i was wondering whether you said that with still the current (hacky/broken) model in mind where we run our chaincode on top of go chaincode (and hence our shim is really just part of the go chaincode and hence was measured in the old model, but seemingly not anymore in the new one, see above)

Reply:
We agree that that part is confusing - whether the peers must agree on the package ID or not.

If yes, I think this solves our problems, without the need of adding MRENCLAVE - it implies they agree on that as well.

If not, of course we have a problem and, yes, we must add MRENCLAVE.
Now, if Fabric enables the customization of the chaincode definition and validation, then let's put it in the chaincode definition.
Otherwise, we should not put it in there, and rather implement a separate mechanism.

> (and hence our shim is really just part of the go chaincode and hence was measured in the old model, but seemingly not anymore in the new one, see above)

hmm I am referring to whatever code/data is reacheable from the Init/Invoke functions which the chaincode must implement and thus must be in the chaincode package that is measured.

Reply:
the package id part is actually clear from doc: the id has only local meaning and does not play any relevance in whether a commit succeeds or not.  From what i've seen from protobufs (protos/peer/lifecycle) seem to confirm that in that there is nothing code related in them such as a package id or alike.

regarding reachable it seems to me you are still in the go-chaincode wrapper model: if we implement our own chaincode type we can do whatever we want in the shim and it doesn't have to be in the chaincode package whatsoever: our chaincode package would just be enclave.so?